passes/hierarchy/hierarchy.cc

```diff
@@ -941,6 +941,61 @@ struct HierarchyPass : public Pass {
 
 		std::set<Module*> blackbox_derivatives;
 		std::vector<Module*> design_modules = design->modules();
+		std::map<Wire*, Cell*> wlogic_map;
+
+		for (auto module : design_modules)
+		for (auto wire : module->wires())
+		{
+			Cell *reduce = nullptr;
+			if (wire->get_bool_attribute("\\wand")) {
+				reduce = module->addCell(
+						stringf("$%s_reduce", wire->name.c_str()), "$reduce_and");
+			}
+			if (wire->get_bool_attribute("\\wor")) {
+				reduce = module->addCell(
+						stringf("$%s_reduce", wire->name.c_str()), "$reduce_or");
+			}
+			if (reduce) {
+				if (wire->width > 1)
+					log_error("Multi-bit wand/wor unsupported (%s)\n",
+							log_id(wire));
+
+				reduce->parameters["\\A_SIGNED"] = Const(0);
+				reduce->parameters["\\A_WIDTH"] = Const(0);
+				reduce->setPort("\\A", SigSpec());
+
+				reduce->parameters["\\Y_WIDTH"] = Const(1);
+				reduce->setPort("\\Y", wire);
+				wlogic_map[wire] = reduce;
+			}
+		}
+
+		for (auto module : design_modules) {
+			std::vector<SigSig> new_connections;
+			for (auto &conn : module->connections())
+			{
+				SigSpec sig = conn.first;
+				for (int i = 0; i < GetSize(sig); i++) {
+					Wire *sigwire = sig[i].wire;
+					if (sigwire == nullptr)
+						continue;
+
+					if (sigwire->get_bool_attribute("\\wor") || sigwire->get_bool_attribute("\\wand")) {
+						Cell *reduce = wlogic_map[sigwire];
+						SigSpec reduce_in = reduce->getPort("\\A");
+						int reduce_width = reduce->getParam("\\A_WIDTH").as_int();
+						Wire *new_reduce_input = module->addWire(
+							stringf("%s_in%d", reduce->name.c_str(), reduce_width));
+						reduce_in.append(new_reduce_input);
+						reduce->setPort("\\A", reduce_in);
+						reduce->fixup_parameters();
+						sig[i] = new_reduce_input;
+					}
+				}
+				new_connections.push_back(SigSig(sig, conn.second));
+			}
+			module->new_connections(new_connections);
+		}
```

Comment:
Is there a better way of doing this than replacing the full connections vector? Couldn't come up with one...

Reply:
Yes, that's the best way to do this.